### PR TITLE
EIP-2537

### PR DESCRIPTION
**Pull request by Marchhill**
## Changes

- Add Ethereum tests for EIP-2537
- Implement fast subgroup checks
- Update precompile addresses and gas costs

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [x] Other: tests

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/Precompiles/Bls/SubgroupChecks.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        Fp2 x = new Fp2(p[..128]).Conjugate() * R;
+        Fp2 y = new Fp2(p[128..]).Conjugate() * S;
+        x.ToBytes(res[..128]);
+        y.ToBytes(res[128..]);
+    }
+
+    internal static BigInteger Normalise(in BigInteger x)
+    {
+        BigInteger unnormalised = x % BlsBaseFieldOrder;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Precompiles/Bls/SubgroupChecks.cs:96]: BigIntegers are painfully slow, do we need simple UInt512 types?

File: src/tests
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/tests:1]: oh I think it is done :)

**Review by MarekM25 — COMMENTED** [review]
don't forget about enabling by timestamp!

**Comment by Marchhill** [comment]
> don't forget about enabling by timestamp!

This was already implemented actually, just had to enable it for Prague

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs [hunk 2]
```diff
@@ -192,7 +192,7 @@ private static void UpdateMetrics(ExecutionOptions opts, UInt256 effectiveGasPri
                 Metrics.BlockMaxGasPrice = Math.Max(gasPrice, Metrics.BlockMaxGasPrice);
 
                 Metrics.AveGasPrice = (Metrics.AveGasPrice * Metrics.Transactions + gasPrice) / (Metrics.Transactions + 1);
-                Metrics.EstMedianGasPrice += Metrics.AveGasPrice * 0.01f * float.Sign(gasPrice - Metrics.EstMedianGasPrice);
+                Metrics.EstMedianGasPrice += Metrics.AveGasPrice * 0.01f * float.Sign(gasPrice - Metrics.AveGasPrice);
```
Comment (Marchhill) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:195]: Had to do this to get the tests to pass, not sure if it is correct

File: src/Nethermind/Nethermind.Evm/Precompiles/Bls/SubgroupChecks.cs (hunk 1 above)
Comment (Marchhill) [reply src/Nethermind/Nethermind.Evm/Precompiles/Bls/SubgroupChecks.cs:96]: At some point the subgroup checks should be implemented in the cryptography library, so probably not worth optimising for now

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs:195]: This shouldn't change.

**Review by MarekM25 — APPROVED** [review]

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**